### PR TITLE
[Merged by Bors] - feat(Algebra): integralClosure as IntermediateField

### PR DESCRIPTION
**Pull request by alreadydone**
+ Add lemmas about integrality and membership in an subalgebra of inverse elements.

+ As a consequence, show the integral closure of a field in a division ring is an intermediate field `IntermediateField.algClosure`.

+ Golf some lemmas along the way.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (alreadydone) [inline Mathlib/RingTheory/Valuation/ValuationSubring.lean:87]: I've been able to show this @MichaelStollBayreuth

File: Mathlib/FieldTheory/Adjoin.lean
Comment (MichaelStollBayreuth) [inline Mathlib/FieldTheory/Adjoin.lean:421]: These porting notes seem to be decidedly unhelpful without knowing what the original proof was...
Can we just remove them?

File: Mathlib/LinearAlgebra/FiniteDimensional.lean
Comment (MichaelStollBayreuth) [inline Mathlib/LinearAlgebra/FiniteDimensional.lean:884]: Why `letI` and not just `let`? (As far as I know, in Lean 4, the `I` is no longer related to instances, but now means "inline".)

File: Mathlib/RingTheory/IntegralClosure.lean
Comment (MichaelStollBayreuth) [inline Mathlib/RingTheory/IntegralClosure.lean:480]:
```suggestion
/-- A commutative domain that is an integral algebra over a field is a field. -/
```

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (MichaelStollBayreuth) [reply Mathlib/RingTheory/Valuation/ValuationSubring.lean:94]: Very nice!

**Comment by MichaelStollBayreuth** [comment]
I've added the `t-algebraic-geometry` label, because this is relevant to the function fields project at the AIM workshop.

**Comment by MichaelStollBayreuth** [comment]
Maybe we can get this into Mathlib before today's problem session at the AIM workshop?

File: Mathlib/LinearAlgebra/FiniteDimensional.lean
Comment (alreadydone) [reply Mathlib/LinearAlgebra/FiniteDimensional.lean:884]:
```suggestion
  let := divisionRingOfFiniteDimensional F K; H.isUnit
```

File: Mathlib/FieldTheory/Adjoin.lean
Comment (alreadydone) [reply Mathlib/FieldTheory/Adjoin.lean:421]: Probably they mean that there exist some shorter proof that times out, so they have more than zero information content ... I'll take a look at the original proof in mathlib3.

**Review by alreadydone — COMMENTED** [review]
Thanks for the quick review, unfortunately I was too busy this morning to answer it before before the afternoon session.

File: Mathlib/FieldTheory/Adjoin.lean
Comment (alreadydone) [reply Mathlib/FieldTheory/Adjoin.lean:421]: Oh, this is actually a proof I golfed. Will remove the porting notes then!

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (alreadydone) [inline Mathlib/RingTheory/Valuation/ValuationSubring.lean:42]: Hi @adamtopaz Do you think (1) this name is good, (2) the argument order is okay (or do you prefer `Place R K`)? This file actually imports TopologicalSpace so we could define the topology on this type, but it's already almost 1000 lines, so maybe put that into another file that constructs the LocallyRingedSpace / Scheme (transc. deg. 1 case) instances.

File: Mathlib/LinearAlgebra/FiniteDimensional.lean
Comment (alreadydone) [inline Mathlib/LinearAlgebra/FiniteDimensional.lean:884]:
```suggestion
  let _ := divisionRingOfFiniteDimensional F K; H.isUnit
```

File: Mathlib/FieldTheory/IntermediateField.lean
Comment (riccardobrasca) [inline Mathlib/FieldTheory/IntermediateField.lean:314]: What about `IntermediateField.integralClosure`? The name `algClosure` looks a little strange.

File: Mathlib/RingTheory/IntegralClosure.lean
Comment (riccardobrasca) [inline Mathlib/RingTheory/IntegralClosure.lean:444]: Can you use at least `variable {R S : Type*}` at the beginning of the section?

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (riccardobrasca) [inline Mathlib/RingTheory/Valuation/ValuationSubring.lean:38]: What about splitting this part to another PR? It seems quite unrelated to the rest.

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/FieldTheory/IntermediateField.lean
Comment (alreadydone) [reply Mathlib/FieldTheory/IntermediateField.lean:314]: It's a bit strange to talk about the integral closure when you know it's a field. There are [AlgebraicClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/IsAlgClosed/AlgebraicClosure.html#AlgebraicClosure) and [IsAlgClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/IsAlgClosed/Basic.html#IsAlgClosure) in mathlib, and I like the shorter name.

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (MichaelStollBayreuth) [reply Mathlib/RingTheory/Valuation/ValuationSubring.lean:38]: This would make sense; perhaps in a new file, maybe `Mathlib.RingTheory.Valuation.Place` (together with the definition of the topology on `Place K R`)? (See the comment on lines +38 to +42.)

**Comment by riccardobrasca** [comment]
Can you please fix the merge conflict? Thanks!

File: Mathlib/RingTheory/Valuation/ValuationSubring.lean
Comment (alreadydone) [reply Mathlib/RingTheory/Valuation/ValuationSubring.lean]: Sorry for the delay; I've removed `Place` from this PR and fixed merge conflicts, and the PR now finally passed CI.

File: Mathlib/RingTheory/IntegralClosure/Algebra/Basic.lean
Comment (alreadydone) [inline Mathlib/RingTheory/IntegralClosure/Algebra/Basic.lean:168]:
```suggestion
/-- The integral closure of `R` in an `R`-algebra `A`. -/
```

**Comment by grunweg** [comment]
This PR looks blocked on review - except that it has a merge conflict, hence does not appear in the review queue...
Can you merge master, please? Thanks!

File: Mathlib/FieldTheory/IntermediateField.lean
Comment (jjdishere) [reply Mathlib/FieldTheory/IntermediateField.lean:270]: In Mathlib, we have `separableClosure` as the relative separable closure (an intermediate field) and `SeperableClosure` as the absolute separable closure. And `normalClosure` for relative normal closure. Do you think `algebraicClosure` is a good name?

Comment (alreadydone) [reply Mathlib/FieldTheory/IntermediateField.lean]: Okay, even though there are IsSepClosed/Closure and IsAlgClosed/Closure, I think we should follow s/SeparableClosure and AlgebraicClosure here to name this algebraicClosure as you said.

File: Mathlib/FieldTheory/IntermediateField/Algebraic.lean
Comment (erdOne) [inline Mathlib/FieldTheory/IntermediateField/Algebraic.lean:40]: I'm not sure if the `algClosure` part is accurate.

**Review by erdOne — APPROVED** [review]
LGTM modulo minor comment

File: Mathlib/FieldTheory/IntermediateField/Algebraic.lean
Comment (alreadydone) [inline Mathlib/FieldTheory/IntermediateField/Algebraic.lean:38]:
```suggestion
/-- The algebraic closure of a field `K` in an extension `L`, the subalgebra `integralClosure K L`
upgraded to an intermediate field (when `K` and `L` are both fields). -/
```

File: Mathlib/FieldTheory/IntermediateField/Algebraic.lean
Comment (alreadydone) [reply Mathlib/FieldTheory/IntermediateField/Algebraic.lean]: Good catch, thanks!

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
